packages/api-server/src/methods/gw-error.ts

```diff
@@ -74,24 +91,108 @@ export function handleGwError(gwJsonRpcError: any) {
+  let extraStack: ExtraStack = ExtraStack.unknown;
+  let extraExitCode: HexNumber = exitCode;
+
+  // if logItem exits, try parse polyjuice/evmc exit code for error message
+  if (logItem) {
+    const polySystemLog = unpackPolyjuiceSystemLog(logItem);
+    const statusCode = polySystemLog.statusCode;
+
+    // 1. parse evmc exit code
+    const evmcCode = matchExitCode(statusCode, EVMC_EXIT_CODE_MAPPING);
+    if (evmcCode != null) {
+      if (evmcCode === EVMC_EXIT_CODE_MAPPING.revert) {
+        // fill the message with detailed abi-encoded revert reason
+        message = parseRevertReason(returnData || "0x");
+      } else {
+        message = evmcCode.message;
+      }
+      extraMessage = evmcCode.message;
+      extraExitCode = exitCodeNumberToHex(statusCode);
+      extraStack = ExtraStack.evmc;
+    }
+
+    // 2. parse polyjuice exit code
+    const polyCode = matchExitCode(statusCode, POLYJUICE_EXIT_CODE_MAPPING);
+    if (polyCode != null) {
+      message = polyCode.message;
+      extraMessage = polyCode.message;
+      extraExitCode = exitCodeNumberToHex(statusCode);
+      extraStack = ExtraStack.poly;
+    }
+
+    // 3. if both un-matched, fallback to parse godwoken exit code
+    if (evmcCode == null && polyCode == null) {
+      const result = parseGwExitCodeForErrorMessage(exitCode);
+      message = result.message;
+      extraMessage = result.extraMessage;
+      extraExitCode = exitCode;
+      extraStack = ExtraStack.gw;
+    }
+  } else {
+    // if logItem not exits, just parse godwoken exit code directly
+    const result = parseGwExitCodeForErrorMessage(exitCode);
+    message = result.message;
+    extraMessage = result.extraMessage;
+    extraExitCode = exitCode;
+    extraStack = ExtraStack.gw;
   }
 
-  // fallback to general revert error
-  handleRevertError(errorTxReceipt);
+  const extra: Extra = {
+    exit_code: extraExitCode,
+    message: extraMessage,
+    stack: extraStack,
+  };
+
+  throw new TransactionExecutionError(message, returnData, extra);
 }
 
-function handleRevertError(errorTxReceipt: ErrorTxReceipt) {
-  const returnData = errorTxReceipt.return_data;
-  const message = parseReturnData(returnData);
-  if (errorTxReceipt.return_data.length > 2) {
-    throw new TransactionExecutionError(message, returnData);
-  } else {
-    throw new TransactionExecutionError(message);
+function parseGwExitCodeForErrorMessage(exitCode: HexNumber): {
+  message: string;
+  extraMessage: string;
+} {
+  let message = "unknown error";
+  let extraMessage = "unknown error";
+  const gwCode = matchExitCode(exitCode, GODWOKEN_EXIT_CODE_MAPPING);
+  if (gwCode != null) {
+    extraMessage = gwCode.message;
+    if (gwCode === GODWOKEN_EXIT_CODE_MAPPING.vmReachedMaxCycles) {
+      message = `out of gas(${gwCode.message})`;
+    } else {
+      message = gwCode.message;
+    }
   }
+  return {
+    message,
+    extraMessage,
+  };
+}
+
+function unpackPolyjuiceSystemLog(logItem: LogItem): PolyjuiceSystemLog {
```

Comment:
https://github.com/godwokenrises/godwoken-web3/blob/c456a58a1126c1bd0d19c0d5142d2a9904fc3422/packages/api-server/src/filter-web3-tx.ts#L266
This function already exists

Reply:
updated